index.js

```diff
@@ -83,6 +94,116 @@ function attachToBody (options, req, reply, next) {
+
+  // first token can't be '[', ']', or '{}'
+  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
+  // only a single {} special ending is allowed
+  else if (tokens.filter(t => t === '{}').length > 1) isValid = false
+  // {}, if present, must be the last token
+  else if (tokens.find(t => t === '{}') && tokens[tokens.length - 1] !== '{}') isValid = false
+
+  let hasOpenBracket = false
+  for (let i = 1; i < tokens.length; i++) {
+    const token = tokens[i]
+    const nextToken = tokens[i + 1]
+    // invalid closing bracket
+    if (!hasOpenBracket && token === ']') {
+      isValid = false
+      break
+    }
+    // invalid opening bracket
+    if (hasOpenBracket && token === '[') {
+      isValid = false
+      break
+    }
+    hasOpenBracket = token === '[' || (hasOpenBracket && token !== ']')
+    // orphan opening bracket
+    if (hasOpenBracket && nextToken === undefined) isValid = false
+    // {} is not allowed within brackets
+    if (hasOpenBracket && token === '{}') isValid = false
+    // after a closing bracket, only an opening bracket or {} are allowed
+    if (token === ']' && nextToken && !['[', '{}'].includes(nextToken)) {
+      isValid = false
+      break
+    }
+  }
+
+  // when the {} special ending is supplied, the value must be valid JSON
+  if (isValid && tokens[tokens.length - 1] === '{}') {
+    try {
+      secureJSON.parse(value)
+    } catch (e) {
+      return 'the {} special ending was used, but value is not a valid JSON'
+    }
+  }
+
+  if (tokens.includes('__proto__')) {
+    return '__proto__ is not allowed as a key name'
+  }
+
+  if (tokens.includes('constructor')) {
+    return 'constructor is not allowed as a key name'
+  }
+
+  return isValid ? null : 'invalid serialized key'
+}
+
+function getSerializedKeySegments (options, body, tokens, value) {
+  const segments = [{ parentRef: body, key: tokens[0] }]
+  const parseArrays = options.enableSpecialNotationKeys !== 'objectsOnly'
+
+  for (let i = 1; i < tokens.length; i++) {
+    const lastSegment = segments[segments.length - 1]
+    const token = tokens[i]
+    const prevToken = tokens[i - 1]
+    let parentRef
+    let key
+
+    if (token === ']' && prevToken === '[') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = parseArrays ? parentRef.length : Object.keys(parentRef).length
+    } else if (parseArrays && token === ']' && isValidIndex(prevToken)) {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = +prevToken
+    } else if (token === ']') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? {}
+      key = prevToken
+    } else if (token === '{}') {
+      value = secureJSON.parse(value)
+    }
+
+    if (parentRef) segments.push({ parentRef, key })
+  }
+
+  segments[segments.length - 1].value = value
+
+  return segments
+}
+
+function isValidIndex (str) {
+  if (str.length > 1 && str[0] === '0') return false
```

Comment:
Maybe a for loop is faster than the regex.

Not tested, maybe wrong. Can  be used for inspiration
```js
const numbersAsciiLookup = Array(256).fill(0).map((v,i) => /^\d+$/.test(String.fromCharCode(i)) && true || false)

function isValidIndex (idx) {
  const len = idx.length
  let result = true
  let i = 0
  let code = 0
  while (i < len) {
    code = idx.charCodeAt(i)
    if (!numbersLookup[code]) {
     result = false
     break;
    }
    if (i === 0 && code === 48 /* "0" */) {
      result = false
      break;
    }
    ++i
  }  
  return result
}
```

Reply:
[I had already benchmarked](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJDaGVja2luZyBpZiBzdHJpbmcgaXMgYSB2YWxpZCBpbnRlZ2VyIG51bWJlciA%2BPSAwIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IFsnMCcsICcwMDEnLCAnMTAwJywgJzEyMzQnLCAnMjInLCAnOTknLCAnYWJjJywgJy0xMjMnLCAnMS43MicsICcxejMnXTtcblxuZnVuY3Rpb24gaXNWYWxpZEluZGV4VjEoc3RyKSB7XG4gIGlmIChzdHIubGVuZ3RoID4gMSAmJiBzdHJbMF0gPT09ICcwJykgcmV0dXJuIGZhbHNlO1xuICByZXR1cm4gL15cXGQrJC8udGVzdChzdHIpO1xufVxuXG5mdW5jdGlvbiBpc1ZhbGlkSW5kZXhWMihzdHIpIHtcbiAgaWYgKHN0ci5sZW5ndGggPiAxICYmIHN0clswXSA9PT0gJzAnKSByZXR1cm4gZmFsc2U7XG4gIGZvciAobGV0IGkgPSAwOyBpIDwgc3RyLmxlbmd0aDsgaSsrKSB7XG4gICAgY29uc3QgYyA9IHN0ci5jaGFyQ29kZUF0KGkpO1xuICAgIGlmIChjIDwgNDggfHwgYyA%2BIDU3KSByZXR1cm4gZmFsc2U7XG4gIH1cbiAgcmV0dXJuIHRydWU7XG59XG5cbmZ1bmN0aW9uIGlzVmFsaWRJbmRleFYzKHN0cikge1xuICBpZiAoc3RyWzBdID09PSAnLScpIHJldHVybiBmYWxzZTtcbiAgcmV0dXJuIHBhcnNlRmxvYXQoc3RyKS50b1N0cmluZygpID09PSBzdHIudG9TdHJpbmcoKTtcbn1cblxuY29uc3QgbnVtYmVyc0FzY2lpTG9va3VwID0gQXJyYXkoMjU2KVxuICAuZmlsbCgwKVxuICAubWFwKCh2LCBpKSA9PiAoL15cXGQrJC8udGVzdChTdHJpbmcuZnJvbUNoYXJDb2RlKGkpKSAmJiB0cnVlKSB8fCBmYWxzZSk7XG5cbmZ1bmN0aW9uIGlzVmFsaWRJbmRleFY0KHN0cikge1xuICBjb25zdCBsZW4gPSBzdHIubGVuZ3RoO1xuICBsZXQgcmVzdWx0ID0gdHJ1ZTtcbiAgbGV0IGkgPSAwO1xuICBsZXQgY29kZSA9IDA7XG4gIHdoaWxlIChpIDwgbGVuKSB7XG4gICAgY29kZSA9IHN0ci5jaGFyQ29kZUF0KGkpO1xuICAgIGlmICghbnVtYmVyc0FzY2lpTG9va3VwW2NvZGVdKSB7XG4gICAgICByZXN1bHQgPSBmYWxzZTtcbiAgICAgIGJyZWFrO1xuICAgIH1cbiAgICBpZiAoaSA9PT0gMCAmJiBsZW4gPiAxICYmIGNvZGUgPT09IDQ4IC8qIFwiMFwiICovKSB7XG4gICAgICByZXN1bHQgPSBmYWxzZTtcbiAgICAgIGJyZWFrO1xuICAgIH1cbiAgICArK2k7XG4gIH1cbiAgcmV0dXJuIHJlc3VsdDtcbn1cblxuIiwidGVzdHMiOlt7Im5hbWUiOiJUZXN0IENhc2UiLCJjb2RlIjoiZGF0YS5mb3JFYWNoKHN0ciA9PiBpc1ZhbGlkSW5kZXhWMShzdHIpKSIsInJ1bnMiOlsxNzUwMDAsMTI0MDAwLDM1NTAwMCwyOTQwMDAsMzYxMDAwLDQyMzAwMCwyNzAwMCwyOTcwMDAsNDAwMCwzMTIwMDAsMTk0MDAwLDIwNjAwMCwzNTUwMDAsMjY0MDAwLDM0NjAwMCwzMzIwMDAsMjIxMDAwLDE1OTAwMCw2OTAwMCwzOTAwMDAsMzU1MDAwLDIwMDAsMzU1MDAwLDMzNDAwMCwxNzMwMDAsODQwMDAsMTM3MDAwLDI3NjAwMCwxMDgwMDAsMjEwMDAwLDIwODAwMCwzNTUwMDAsMTAwMCwxNzIwMDAsMzg5MDAwLDg1MDAwLDUwMDAwLDIxMTAwMCw0MzIwMDAsODcwMDAsNTYwMDAsMTMxMDAwLDExOTAwMCwzNTgwMDAsMTQyMDAwLDE1MDAwLDIwMjAwMCw0MDMwMDAsMTAxMDAwLDIyODAwMCwzNTUwMDAsNDUyMDAwLDM5MDAwLDE4MzAwMCwyNDkwMDAsNDQ4MDAwLDM1NTAwMCw1MTAwMCwyNDcwMDAsMTE5MDAwLDMwODAwMCwyNTgwMDAsOTQwMDAsMTU2MDAwLDExMDAwLDIxMTAwMCwzMTEwMDAsNDE1MDAwLDQxNjAwMCwzMTgwMDAsODAwMDAsMjI4MDAwLDIzMTAwMCwzNTUwMDAsMjIwMDAsMzU1MDAwLDEwMDAsMzU1MDAwLDkzMDAwLDMxODAwMCwzMTIwMDAsMzI3MDAwLDEwMDAsMjk4MDAwLDI0NDAwMCwzNTUwMDAsMjcwMDAsNzMwMDAsMjA5MDAwLDM1NTAwMCwxNDYwMDAsMzU1MDAwLDI0MzAwMCw5OTAwMCwyNTEwMDAsMTQ5MDAwLDE2NzAwMCwxNTcwMDAsMzU1MDAwLDgyMDAwXSwib3BzIjoyMTkyNjB9LHsibmFtZSI6IlJlZ2V4LWJhc2VkIiwiY29kZSI6ImRhdGEuZm9yRWFjaChzdHIgPT4gaXNWYWxpZEluZGV4VjIoc3RyKSkiLCJydW5zIjpbNDgwMDAsNjMwMDAsMjYwMDAwLDE5ODAwMCwyOTIwMDAsMjczMDAwLDMwMDAwLDIwMTAwMCwxMDAwLDIzMjAwMCwxMzIwMDAsMTM1MDAwLDI3NzAwMCwxNTQwMDAsMTk0MDAwLDIzOTAwMCwxNDcwMDAsMTI2MDAwLDUwMDAwLDEzMjAwMCwyNjMwMDAsMTAwMCwyMjIwMDAsMjYwMDAsMjUwMDAsMTU3MDAwLDEwNzAwMCwxMDAwLDE5ODAwMCwxNjIwMDAsMTIwMDAwLDI2MTAwMCwyMTAwMCwyOTAwMCw3NDAwMCwxMDMwMDAsMzEwMDAwLDEzMjAwMCwzMzEwMDAsMzQwMDAsMzQxMDAwLDEwMDAwLDIzNzAwMCwyODIwMDAsMTY5MDAwLDE0MjAwMCwyMDAwLDI3NDAwMCwzMjgwMDAsMjM0MDAwLDIzNjAwMCwzNTkwMDAsMTA1MDAwLDI5MzAwMCwxODIwMDAsMjg2MDAwLDE1MzAwMCwyMTAwMCwxMjgwMDAsMjE1MDAwLDIxNDAwMCwxMzIwMDAsNzMwMDAsMTAwMCw3MDAwLDExNTAwMCwxNzIwMDAsMzE1MDAwLDE2MTAwMCwyOTcwMDAsMTY0MDAwLDI4MjAwMCwyMTEwMDAsMTk0MDAwLDE1NTAwMCwzNTAwMCwxMDAwLDEzMjAwMCw4NjAwMCwyMjMwMDAsMjUzMDAwLDEzMjAwMCwyODgwMDAsMTYxMDAwLDE4MDAwMCwyMzMwMDAsMjIwMDAsMTAwMCwxMzIwMDAsMjc2MDAwLDExODAwMCwzMTAwMDAsMjE0MDAwLDI0ODAwMCwxMzIwMDAsMTA1MDAwLDEzMjAwMCwxMjIwMDAsMjQxMDAwLDY1MDAwXSwib3BzIjoxNjA5MzB9LHsibmFtZSI6Ikl0ZXJhdGlvbi1iYXNlZCIsImNvZGUiOiJkYXRhLmZvckVhY2goc3RyID0%2BIGlzVmFsaWRJbmRleFYzKHN0cikpIiwicnVucyI6WzIwMDAsMzIwMDAsMjIyMDAwLDE5NjAwMCwzMzEwMDAsMzAxMDAwLDEwMDAsMTg4MDAwLDMzMTAwMCwxMTAwMDAsMzA4MDAwLDE1NDAwMCwyNTEwMDAsMTAwMDAwLDUyMDAwLDIwMDAwMCw5OTAwMCw1OTAwMCwxMjAwMCwzMjcwMDAsMjg2MDAwLDM2MzAwMCwyNTEwMDAsMTAwMCwzNjMwMDAsMTY1MDAwLDI3ODAwMCwzMzEwMDAsMjAwMDAwLDUxMDAwLDIzNjAwMCwxMDAwLDEwMDAsMTIzMDAwLDMxNzAwMCw1MzAwMCwzMTQwMDAsMTI1MDAwLDMxNTAwMCwyMjAwMCwzMjcwMDAsMTAwMCwyNzQwMDAsNDkwMDAsMTczMDAwLDI4NTAwMCwxMDAwLDExMjAwMCwyODAwMCwxNDUwMDAsMjQ4MDAwLDE3MzAwMCw1MDAwMCwxMDkwMDAsMzYwMDAwLDI1MzAwMCwxMDAwLDE0MTAwMCwxMDAwLDY1MDAwLDEwMDAsNjYwMDAsMTAwMCwxMDAwLDIwODAwMCwxOTIwMDAsMjQ2MDAwLDMzMTAwMCwyNzkwMDAsMTg1MDAwLDI2NjAwMCwzMzEwMDAsMzMxMDAwLDEzNTAwMCwxMzQwMDAsMTAwMCwyMzgwMDAsNjYwMDAsMTgwMDAsMTAwMCwxMTUwMDAsNzIwMDAsMzMxMDAwLDE0MzAwMCwxODcwMDAsMjM5MDAwLDEwMDAsMTAwMCwxMTAwMDAsMjgzMDAwLDI0MTAwMCwxMzYwMDAsMTUwMDAwLDIyODAwMCwxMTgwMDAsOTAwMDAsNjQwMDAsMzQ1MDAwLDMyMzAwMCwxMDAwXSwib3BzIjoxNjA3NzB9LHsibmFtZSI6InBhcnNlRmxvYXQtYmFzZWQiLCJjb2RlIjoiZGF0YS5mb3JFYWNoKHN0ciA9PiBpc1ZhbGlkSW5kZXhWNChzdHIpKSIsInJ1bnMiOlsyMjMwMDAsMjU4MDAwLDg4MDAwLDE5MDAwLDE3MDAwMCwyNzcwMDAsMjA3MDAwLDE4ODAwMCwzMjMwMDAsMTg0MDAwLDI0MTAwMCwzMDAwLDIyMzAwMCwxNDAwMDAsNzkwMDAsMjE3MDAwLDkzMDAwLDc2MDAwLDIxMDAwMCwxODgwMDAsMjkyMDAwLDE5ODAwMCw0MzAwMCwxMTEwMDAsMjE1MDAwLDEwMDAsMjAyMDAwLDE2NzAwMCwxNzIwMDAsMzUxMDAwLDE2NTAwMCwyNjAwMCwxOTEwMDAsMjI3MDAwLDY0MDAwLDI4MjAwMCwxMDEwMDAsNjMwMDAsMTk2MDAwLDE4ODAwMCwyMTkwMDAsMTAwMCwxNTkwMDAsMTgyMDAwLDExNzAwMCwxNjQwMDAsNDAwMDAsMjM4MDAwLDEyMDAwMCwxOTEwMDAsMjUwMDAsMTE5MDAwLDI4NzAwMCwxODgwMDAsMTU3MDAwLDE1OTAwMCwzMzkwMDAsMTg4MDAwLDUwMDAsMTU5MDAwLDMxMzAwMCwyODQwMDAsMTAwMCw1MjAwMCwxMDAwLDE2NjAwMCwxMjIwMDAsMzQwMDAsMTE3MDAwLDkwMDAsMjMzMDAwLDEyNjAwMCwyMzYwMDAsNTIwMDAsMzIxMDAwLDM1MzAwMCwxMDAwLDY3MDAwLDQ5MDAwLDk4MDAwLDE4NTAwMCwxMjAwMCwyNzAwMDAsMjEwMDAsMTQwMDAwLDE4MzAwMCwyNTAwMDAsMzU2MDAwLDExMDAwMCwxNTAwMDAsMTYyMDAwLDI4NDAwMCw1OTAwMCw2MTAwMCwxODgwMDAsMzE1MDAwLDkyMDAwLDI2OTAwMCwxNjAwMDAsMjUzMDAwXSwib3BzIjoxNTg3NDB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMTFUMTk6NTI6MjMuNjMwWiJ9) the regex-version against other two alternatives. Regex had won. I included your suggestion as a 4th alternative, but regex still wins, so I'm keeping it.

Reply:
Usually benchmarking is done in node and not in browser. 

But check [this](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJDaGVja2luZyBpZiBzdHJpbmcgaXMgYSB2YWxpZCBpbnRlZ2VyIG51bWJlciA%2BPSAwIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IFsnMCcsICcwMDEnLCAnMTAwJywgJzEyMzQnLCAnMjInLCAnOTknLCAnYWJjJywgJy0xMjMnLCAnMS43MicsICcxejMnXTtcblxuZnVuY3Rpb24gaXNWYWxpZEluZGV4VjEoc3RyKSB7XG4gIGlmIChzdHIubGVuZ3RoID4gMSAmJiBzdHJbMF0gPT09ICcwJykgcmV0dXJuIGZhbHNlO1xuICByZXR1cm4gL15cXGQrJC8udGVzdChzdHIpO1xufVxuXG5jb25zdCBtdWx0aURpZ2l0TnVtYmVyUkUgPSAvXlxcZCR8XlteMF1cXGQrJC9cbmNvbnN0IGlzVmFsaWRJbmRleFY1ID0gbXVsdGlEaWdpdE51bWJlclJFLnRlc3QuYmluZChtdWx0aURpZ2l0TnVtYmVyUkUpXG4iLCJ0ZXN0cyI6W3sibmFtZSI6IlRlc3QgQ2FzZSIsImNvZGUiOiJkYXRhLmZvckVhY2goc3RyID0%2BIGlzVmFsaWRJbmRleFYxKHN0cikpIiwicnVucyI6WzExMzAwMCwzMTEwMDAsMTAwMCwyMDAwLDEwMDAsNTQwMDAsNjUwMDAsNzIwMDAsODEwMDAsMjcwMDAwLDE1MTAwMCwzMTcwMDAsMTU4MDAwLDEwMDAsMzE3MDAwLDE1NjAwMCwyOTAwMCw5OTAwMCwxMDAwLDIzMDAwLDU5MDAwLDMyMDAwMCwxMDAwLDMwMTAwMCw3OTAwMCwzOTAwMCw1MTAwMCwxMDAwLDU5MDAwLDEwMDAsMTc1MDAwLDYyMDAwLDEwNjAwMCw4MzAwMCwxMDAwLDI2MzAwMCwzMTAwMCwxMDAwLDIyMDAwLDE1MDAwLDE4MDAwMCwxNTgwMDAsMzEwMDAsMTMzMDAwLDExMzAwMCwxMDAwLDcwMDAwLDE4NjAwMCwxNDAwMDAsMTYwMDAsNTYwMDAsOTUwMDAsNjcwMDAsMjkwMDAwLDExNTAwMCwxMDAwLDI1NzAwMCwyMjYwMDAsMTM2MDAwLDkxMDAwLDk2MDAwLDYwMDAwLDEwMDAsMTI0MDAwLDEwMDAsNDAwMCwxMDYwMDAsMjUzMDAwLDEwOTAwMCwzNjAwMCwzMTgwMDAsMjg0MDAwLDE4ODAwMCwxMDAwLDEwMDAsMTU1MDAwLDEwMDAsOTcwMDAsMTQ5MDAwLDIzNDAwMCw1MzAwMCwxNjQwMDAsMTQ5MDAwLDIyMDAwMCwxMDAwLDE3MTAwMCwxMDAwLDQwMDAwLDEyNzAwMCw4NjAwMCwzNDMwMDAsNDkwMDAsMTc1MDAwLDE3MTAwMCw3NzAwMCwyNTMwMDAsMjE0MDAwLDEyMjAwMCwxMzAwMDAsMzIwMDBdLCJvcHMiOjExMDUwMH0seyJuYW1lIjoiVGVzdCBDYXNlIiwiY29kZSI6ImRhdGEuZm9yRWFjaChzdHIgPT4gaXNWYWxpZEluZGV4VjUoc3RyKSkiLCJydW5zIjpbMzE0MDAwLDU3MDAwLDYwMDAsMTk4MDAwLDMyMzAwMCwyODMwMDAsMTQ1MDAwLDM4OTAwMCw0NzAwMCwyNDcwMDAsMjA1MDAwLDQwMDAwLDE4NzAwMCwxMjUwMDAsMjM5MDAwLDUwNTAwMCw0NDAwMCwyMjgwMDAsNDc0MDAwLDEwNDAwMCw5NTAwMCw0NjcwMDAsNTQwMDAsMTA0MDAwLDM1MzAwMCw1NjUwMDAsMjY1MDAwLDEwMDAsMjA5MDAwLDE5ODAwMCwxMDAwLDE5OTAwMCwxNzgwMDAsNTUwMDAsMjI5MDAwLDU3MDAwLDE0NjAwMCwyMzQwMDAsMjkyMDAwLDkyMDAwLDE5MDAwLDExMzAwMCw0MTAwMCwxMzkwMDAsMzg5MDAwLDM1NTAwMCw0MjAwMCw1MjAwMCw5MDAwLDI0MDAwMCwxMTMwMDAsMjkwMDAsMjQ4MDAwLDEwMDAsMjc3MDAwLDIwMDAsMTAwMCwwLDI5MDAwMCwxMDAwLDMwNzAwMCwxMDAwLDEwMDAsMzMwMDAsMzA2MDAwLDQ2NDAwMCw0MTUwMDAsMTg5MDAwLDExOTAwMCwxMDAwLDE0NzAwMCwxNjcwMDAsNTE4MDAwLDQwMTAwMCwxMjIwMDAsMzA4MDAwLDIwMTAwMCwxMDAwLDgwMDAwLDM5MDAwLDI5OTAwMCwxMDAwLDE2NzAwMCwyNTIwMDAsMzY2MDAwLDI0NjAwMCwyMjgwMDAsODQwMDAsNDMwMDAsMTg0MDAwLDEwMDAsMjE2MDAwLDQ3MDAwLDE3MDAwLDE3NjAwMCwxMDAwLDEwNzAwMCwxODAwMCwxMzQwMDAsNDM1MDAwXSwib3BzIjoxNzE1NzB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMTFUMjM6MTM6MzEuNzE4WiJ9) out:

Reply:
Your version allows any character as the first one, except for 0. For example, `#123`, `)2222` and `*1` are all valid.

Reply:
how about ^\d$|^[1-9]\d+$

Reply:
This one works as expected and performs slightly better than the previous one. I've changed it. 👍🏼